File: rec_des_par.py
i = 0
t = 1
str_input = input("Enter the string to be parsed: ")

def E():
    T()
    EPRIME()

def EPRIME():
    global i
    if i < len(str_input) and str_input[i] == '+':
        i += 1
        T()
        EPRIME()

def T():
    F()
    TPRIME()

def TPRIME():
    global i
    if i < len(str_input) and str_input[i] == '*':
        i += 1
        F()
        TPRIME()

def F():
    global i, t
    if i < len(str_input) and str_input[i] == 'a':
        i += 1
    elif i < len(str_input) and str_input[i] == '(':
        i += 1
        E()
        if i < len(str_input) and str_input[i] == ')':
            i += 1
        else:
            error()
    else:
        error()

def error():
    global t
    t = -1

E()

if t == 1 and i == len(str_input):
    print("The given string is accepted")
else:
    print("The given string is not accepted")

File: rec_des_par.c
#include <stdio.h>
#include <stdlib.h>

int i = 0, t = 1;
char str[20];

void E();
void EPRIME();
void T();
void TPRIME();
void F();
void error();

int main() {
    printf("The given grammar is:\n");
    printf("\nE -> TE'");
    printf("\nE' -> +TE'/$");
    printf("\nT -> FT'");
    printf("\nT' -> *FT'/$");
    printf("\nF -> (E)/a");
    
    printf("\nEnter the string to be parsed: ");
    scanf("%s", str);
    
    E();
    
    if (t != 1 || str[i] != '\0') {
        printf("\nGiven string is not accepted\n");
    } else {
        printf("\nThe given string is accepted\n");
    }

    return 0;
}

void E() {
    T();
    EPRIME();
}

void EPRIME() {
    if (str[i] == '+') {
        i++;
        T();
        EPRIME();
    }
}

void T() {
    F();
    TPRIME();
}

void TPRIME() {
    if (str[i] == '*') {
        i++;
        F();
        TPRIME();
    }
}

void F() {
    if (str[i] == 'a') {
        i++;
    } else if (str[i] == '(') {
        i++;
        E();
        if (str[i] == ')') {
            i++;
        }
    } else {
        error();
    }
}

void error() {
    t = -1;
}


OUTPUT

The given grammar is:

E -> TE'
E' -> +TE'/$
T -> FT'
T' -> *FT'/$
F -> (E)/a
Enter the string to be parsed: a+a*a

The given string is accepted
